### ISSUE TITLE
Epic: Implement @skroyc/rmm-middleware - Reflective Memory Management for LangChain

### ISSUE THREAD

**Issue by SkrOYC** [SkrOYC/langchain-middlewares-callbacks-ts#43]
# Epic: Implement @skroyc/rmm-middleware - Reflective Memory Management for LangChain

## Overview

This epic tracks the implementation of the **Reflective Memory Management (RMM)** middleware package for LangChain, based on the paper "In Prospect and Retrospect: Reflective Memory Management for Long-term Personalized Dialogue Agents" (Tan et al., ACL 2025, arXiv:2503.08026v2).

The RMM framework addresses critical limitations in current LLM-based dialogue agents by enabling:
- **Long-term memory retention** across multiple sessions
- **Adaptive retrieval** that learns from user-specific patterns without labeled data
- **Dynamic memory organization** at natural semantic boundaries

## Architecture Summary

RMM operates within LangChain's Middleware API layer using five lifecycle hooks mapped to Algorithm 1 from the paper:

```
beforeAgent  → beforeModel  → wrapModelCall → afterModel → afterAgent
(Load)         (Retrieve)    (Rerank+Inject)   (RL Update)  (Extract)
```

### Three-Layer Storage Architecture

| Layer | Scope | System | Data Stored |
|-------|-------|--------|-------------|
| **Conversation** | Thread | State (Checkpointer) | Session markers, turn counters |
| **Memory Content** | User | VectorStore | Memory documents + embeddings |
| **RL State** | User | BaseStore | Reranker weights (W_q, W_m), metadata |

### Simplified Session Model

Unlike the paper's explicit session boundaries, this implementation uses a simplified model:
- **No explicit `isSessionEnd` signal** - Prospective Reflection runs after every `afterAgent`
- Each thread is treated as a continuous session
- Memory extraction happens automatically at the end of each agent invocation

### Equation 1: Embedding Adaptation
```
q' = q + W_q · q
m'_i = m_i + W_m · m_i
```

### Equation 2: Gumbel-Softmax Sampling
```
g_i = -log(-log(u_i)),  u_i ~ Uniform(0,1)
s̃_i = s_i + g_i
p_i = exp(s̃_i/τ) / Σ_j exp(s̃_j/τ)
```

### Equation 3: REINFORCE Update
```
Δφ = η · (R - b) · ∇_φ log P(M_M | q, M_K; φ)
```

## Implementation Approach

- **Test-Driven Development (TDD)**: Each sub-issue follows RED-GREEN-REFACTOR cycle
- **Testing Framework**: Bun:test only (no Jest/Vitest)
- **Mocking**: All external dependencies (LLM, VectorStore, BaseStore) mocked in tests
- **Build System**: tsup with ESM-only output
- **Type Safety**: TypeScript strict mode, Zod schemas for all data
- **Error Handling**: Fail-safe degradation (abort RMM on any error, return empty delta)

## Hyperparameters (Paper Defaults)

| Parameter | Value | Description |
|-----------|-------|-------------|
| topK | 20 | Retrieved candidates |
| topM | 5 | Reranked candidates sent to LLM |
| temperature (τ) | 0.5 | Gumbel-Softmax temperature |
| learning rate (η) | 1e-3 | REINFORCE learning rate |
| baseline (b) | 0.5 | REINFORCE variance reduction |
| batch size | 4 | Gradient accumulation turns |

## Sub-Issues

1. **#44** - [RMM-1] Project scaffolding and Zod schemas for data models
2. **#45** - [RMM-2] Core utilities - matrix operations, cosine similarity, citation extractor
3. **#46** - [RMM-3] Storage adapters for BaseStore (weights and metadata)
4. **#47** - [RMM-4] Built-in prompts for Prospective Reflection (Appendix D.1.1, D.1.2)
5. **#48** - [RMM-5] Prospective Reflection implementation - extraction, merge/add, afterAgent hook
6. **#49** - [RMM-6] Retrospective Reflection - Retrieval (beforeAgent, beforeModel, wrapModelCall)
7. **#50** - [RMM-7] Retrospective Reflection - Learning (afterModel hook with REINFORCE)
8. **#51** - [RMM-8] Production hardening - fail-safe error handling and documentation

## Technical Constraints

- **CPU-Only Math**: No GPU acceleration (reranker uses simple matrix operations, fast enough on CPU)
- **Pure CPU Implementation**: ~18MB weight matrices (2 × 1536 × 1536 floats), updates in <1ms
- **InMemoryVectorStore**: For testing (production can use any VectorStoreInterface)
- **Bun:test Only**: No other testing frameworks
- **Monorepo Compatibility**: Follow same patterns as `@skroyc/ag-ui-middleware-callbacks`

## Definition of Done

- [ ] All 8 sub-issues closed
- [ ] 100% test coverage for all algorithms
- [ ] All hooks integrated and tested
- [ ] Error handling verified for all failure modes
- [ ] Documentation complete (README, API docs, examples)
- [ ] Package builds successfully with tsup
- [ ] TypeScript strict mode passes
- [ ] Ready for npm publish

## References

- **SPEC.md**: `/packages/rmm-middleware/SPEC.md` (1313 lines, comprehensive specification)
- **Paper**: arXiv:2503.08026v2 (ACL 2025)
- **LangChain Middleware API**: Uses `createMiddleware()` from `langchain`
- **Related Packages**: `@skroyc/ag-ui-middleware-callbacks` (pattern reference)
